1.0 upgrade guide restructure (#1914)
DataDog/dd-trace-rb: docs/UpgradeGuide.md

# From 0.x to 1.0

> Upgrading `ddtrace` from 0.x to 1.x introduces some changes to namespacing, the public API, and the underlying trace data structure.

Author: I like the general idea (expanding upon expectations), but I would expect a more formal/impersonal tone for such documentation. I may take the broad idea here and rework it.

> **How to upgrade basic usage** *(Out-of-the-box instrumentation)*

Reviewer: My view is the reverse: I wouldn't call turning on all products and having a lot of configuration "basic". I'm willing to disagree and commit.

  - [Activating instrumentation](#1.0-configuration-instrumentation)
- [Instrumentation](#1.0-instrumentation)
  - [Service naming](#1.0-instrumentation-service-naming)

**Addtional upgrades for advanced usage** *(Manual instrumentation)*

- [Namespacing](#1.0-namespacing)
- [Trace API](#1.0-trace-api)
  - [Removed `Datadog.tracer`](#1.0-trace-api-removed-tracer)
  - [Removed access to `Datadog::Context`](#1.0-trace-api-removed-context)
  - [Manual tracing & trace model](#1.0-trace-api-manual-tracing)
  - [Accessing trace state](#1.0-trace-api-trace-state)
  - [Distributed tracing](#1.0-trace-api-distributed)
    - [Over HTTP](#1.0-trace-api-distributed-http)
    - [Over gRPC](#1.0-trace-api-distributed-grpc)
    - [Between threads](#1.0-trace-api-distributed-threads)
  - [Sampling](#1.0-trace-api-sampling)
  - [Processing pipeline](#1.0-trace-api-pipeline)

**Appendix**

  - [Namespace mappings](#1.0-appendix-namespace)
    - [`require` paths](#1.0-appendix-namespace-requires)
    - [Constants](#1.0-appendix-namespace-constants)
  - [Breaking changes](#1.0-appendix-breaking-changes)

<h1 id="1.0-basic-upgrade">Upgrading basic usage</h1>

> 

Reviewer: I touched a bit on this on my suggestions to the intro, but I think we should clarify at the beginning of this section what we mean by basic[/standard] usage, because while it's quite clear to us the developers which one is which, customers will probably not know on which bucket they are.

Here's my suggestion for that intro. It repeats a bit what I suggested we say above, but  makes more explicit what we mean by "standard" usage.

> For most users, moving to 1.0 is as simple as changing a few (possibly no lines of code).
>
> If you're only using automatic instrumentation and don't use `Datadog.configure` to change the gem, then you're done, no changes are needed! 🎉 
>
> If you're configuring `ddtrace` via `Datadog.configure`, read on for a few changes that you'll need to do.
> 
> If, additionally, you're calling directly into `ddtrace` APIs from your code -- anything beyond `Datadog.configure`, you should additionally look to the _Additional upgrades for advanced usage_ section.

Perhaps we could even add:

> Did we miss something? Please do [get in contact](https://github.com/DataDog/dd-trace-rb/issues) -- we want to ensure your upgrade experience is brilliant!

Author: I agree it would be better to clarify what qualifies a user for each group. However, I think this description belongs at the top, and (again) I don't think we should use such a personal or casual tone.

# From 0.x to 1.0

Upgrading `ddtrace` from 0.x to 1.x introduces some breaking changes which are outlined below.

**How to upgrade basic usage**

> For users with a basic implementation (configuration file + out-of-the-box instrumentation), only minor changes to your configuration file are required: most applications take just minutes to update. Check out the following sections for a step-by-step guide.

Reviewer: I'm still quite unconvinced with the use of the word _basic_. Since you didn't like my suggestion of _standard_, I've gotten my thesaurus out and here's a few more suggestions: base, common, simple, typical, regular, ~bourgeois~, ~philistine~.

All of them I think don't carry the weird connotations that sometimes get attached to _basic_ 😄

- [Configuration](#1.0-configuration)
  - [`require` paths have changed](#1.0-configuration-requires)
  - [Settings have been renamed](#1.0-configuration-settings)
  - [Activating instrumentation](#1.0-configuration-instrumentation)
- [Instrumentation](#1.0-instrumentation)
  - [Service naming](#1.0-instrumentation-service-naming)

> **Addtional upgrades for advanced usage**

Reviewer: ```suggestion
**Additional upgrades for advanced usage**
```

> For users with an advanced implementation of `ddtrace` (custom instrumentation, sampling or processing behavior, etc), some namespace and behavioral changes may be required. See the following sections for details about what features changed and how to use them.
> 

Reviewer: I guess it makes sense to underline the "additionalness" of these steps again here:

```suggestion
For users with an advanced implementation of `ddtrace` (custom instrumentation, sampling or processing behavior, etc), some namespace and behavioral changes may be additionally required. On top of the changes listed for base usage, see the following sections for details about what features changed and how to use them.
```

- [Namespacing](#1.0-namespacing)
- [Trace API](#1.0-trace-api)
  - [Removed `Datadog.tracer`](#1.0-trace-api-removed-tracer)
  - [Removed access to `Datadog::Context`](#1.0-trace-api-removed-context)
  - [Manual tracing & trace model](#1.0-trace-api-manual-tracing)
  - [Accessing trace state](#1.0-trace-api-trace-state)
  - [Distributed tracing](#1.0-trace-api-distributed)
    - [Over HTTP](#1.0-trace-api-distributed-http)
    - [Over gRPC](#1.0-trace-api-distributed-grpc)
    - [Between threads](#1.0-trace-api-distributed-threads)
  - [Sampling](#1.0-trace-api-sampling)
  - [Processing pipeline](#1.0-trace-api-pipeline)

**Appendix**

For a comprehensive list of everything that changed, the appendix hosts some helpful and detailed tables with recommendations.

  - [Namespace mappings](#1.0-appendix-namespace)
    - [Constants](#1.0-appendix-namespace-constants)
  - [Breaking changes](#1.0-appendix-breaking-changes)

> <h1 id="1.0-basic-upgrade">Upgrading basic usage</h1>

Reviewer: It's a bit weird that while this document (`UpgradeGuide.md`) seems like it's making space for future upgrades to be included as well, most sections are using h1's, aka putting stuff on the top-level structure.

Should we make this document just the 1.x upgrade guide? In that case, it makes sense to keep the h1s. Otherwise, should everything below "From 0.x to 1.0" become an h2?

  c.agent.hostname = '127.0.0.1'
  c.agent.port = 8126

>   c.diagnostics.debug = true

Author: I'll take a look at removing some of these.

>   c.diagnostics.debug = true

Author: I disagree: although I can understand why it would be appealing to keep ordering the same logically, visual appeal of the outcome is compromised if sensible groups/alphabetical order isn't maintained when converted.

>   c.diagnostics.debug = true

Reviewer: I think we're in alignment. My suggestion was not to break the new groups, but to reorder things in the new group in the same order as they appeared in the 0.x config:

```ruby
  # Global settings
  c.diagnostics.debug = true
  c.service = 'billing-api'
  c.runtime_metrics.enabled = true
  c.agent.hostname = '127.0.0.1'
  c.agent.port = 8126
```

end
```

> Check out the table below for a list of common mappings:

Author: It isn't a full list, its an abbreviated list. And the implication behind common is correct; it may be missing some mappings, which we excluded for brevity (e.g. `tracing.distributed_tracing` vs `tracing.distributed_tracing.propagation_extract_style` and `tracing.distributed_tracing.propagation_inject_style`)

| `tracer.instance`                      | `tracing.instance`            |
| `tracer.partial_flush`                 | `tracing.partial_flush`       |
| `tracer.port`                          | `agent.port`                  |
| `tracer.sampler`                       | `tracing.sampler`             |
| `tracer.transport_options`             | `tracing.transport_options`   |
| `tracer.transport_options(options)`    | Removed                       |
| `tracer.writer`                        | `tracing.writer`              |
| `tracer.writer_options`                | `tracing.writer_options`      |

<h3 id="1.0-configuration-instrumentation">Activating instrumentation</h3>

The `use` function has been renamed to `instrument`. Also, `instrument` has been namespaced to the specific feature:

```ruby
### Old 0.x ###
Datadog.configure do |c|
  # Tracing instrumentation
  c.use :rails

  # CI instrumentation
  c.use :cucumber
end

### New 1.0 ###
Datadog.configure do |c|
  # Tracing instrumentation
  c.tracing.instrument :rails

  # CI instrumentation
  c.ci.instrument :cucumber
end
```

<h2 id="1.0-instrumentation">Instrumentation</h2>

> <h3 id="1.0-instrumentation-service-naming">Service naming</h3>

Reviewer: Hmm, could we compromise and have a note here about renaming, and push the full detail with screenshots to the appendix?

> <h3 id="1.0-instrumentation-service-naming">Service naming</h3>

Author: While that would help brevity, it creates a situation where jumping between two parts of the document to complete a story disrupts comprehension and introduces some confusion in navigation.

I think the appendix best embodies things like huge tables that are pure reference material, but these screenshots are constitutional to the "service name change" section. Moving it to the appendix would create a disjointed reading experience.

| `Datadog.tracer.set_tags(tags)`                                                              | `Datadog.configure { \|c\| c.tags = tags }` |                                                                     |
| `Datadog.tracer.shutdown!`                                                                   | `Datadog::Tracing.shutdown!`                |                                                                     |
| `Datadog.tracer.start_span`                                                                  | `Datadog::Tracing.trace`                    | See [manual tracing](#1.0-trace-api-manual-tracing) for details.    |
| `Datadog.tracer.trace`                                                                       | `Datadog::Tracing.trace`                    | See [manual tracing](#1.0-trace-api-manual-tracing) for details.    |

Also check out the functions defined within `Datadog::Tracing` in our [public API](https://www.rubydoc.info/gems/ddtrace/) for more details on their usage.

<h3 id="1.0-trace-api-removed-tracer">Removed `Datadog.tracer`</h3>

Many of the functions accessed directly through `Datadog.tracer` have been moved to `Datadog::Tracing` instead.

> <h3 id="1.0-trace-api-removed-context">Removed access to `Datadog::Context`</h3>

Author: This is expected. Markdown does not support custom IDs for headers: it will always use the header's name as an ID. This will inevitably cause conflict if there are subsections bearing the same name.

> <h3 id="1.0-trace-api-removed-context">Removed access to `Datadog::Context`</h3>

Reviewer: Indeed. My comment was wondering if it made sense to remove the markdown (the backquotes) in this case, since it gets rendered literally.

> <h3 id="1.0-trace-api-removed-context">Removed access to `Datadog::Context`</h3>

Author: Which I think is fine: helps make the keyword clear, even if it doesn't apply the beautiful styling you would normally expect.

Direct usage of `Datadog::Context` has been removed. Previously, it was used to modify or access active trace state. Most use cases have been replaced by our public trace API.

<h3 id="1.0-trace-api-manual-tracing">Manual tracing & trace model</h3>

> Manual tracing is now done through the public API.

Reviewer: Could you add a link to "public API" here?

> 

Reviewer: ⬇️ I can't comment on the next line directly (github...) but I found it confusing, since it states that "the block would yield" where actually it's the `Tracer#trace` method that yields to the block, and not the reverse.

Author: I don't see the confusion here, as a block yields an argument, even as a function yields arguments to a block.

Author: I think you may be getting a bit hung up on the special usage of the word yield in Ruby, whereas [in normal English](https://dictionary.cambridge.org/us/dictionary/english/yield) it means "provide", which is how I was using it here. I've changed "yield" to "provide" which should hopefully clarify and resolve this.

Reviewer: In my defense, if I can get confused by this, our customers can too ;)

Author: Yup, hence the change to "provide" instead.

Reviewer: Hmm I've reviewed the change and I still think it's not as clear is it could be. It's now "the block would provide" and "the block provides", where it's actually the reverse: the block gets provided/called -- it's on the receiving end, not on the supplying end.

![1.0 trace](./1.0-trace.png)

<h1 id="1.0-advanced-upgrade">Upgrading advanced usage</h1>

<h2 id="1.0-namespacing">Namespacing</h2>

Many files and constants within `ddtrace` have been recategorized by feature. The new categorization scheme is as follows:

| Feature          | Namespace                | File path               |
|------------------|--------------------------|-------------------------|
| Globals          | `Datadog`                | `ddtrace`               |
|                  |                          |                         |
| CI               | `Datadog::CI`            | `datadog/ci`            |
| Core (Internals) | `Datadog::Core`          | `datadog/core`          |

> | OpenTelemetry    | `Datadog::OpenTelemetry` | `datadog/opentelemetry` |

Reviewer: Should be removed :)

| `Datadog.tracer.start_span`                                                                  | `Datadog::Tracing.trace`                    | See [manual tracing](#1.0-trace-api-manual-tracing) for details.    |
| `Datadog.tracer.trace`                                                                       | `Datadog::Tracing.trace`                    | See [manual tracing](#1.0-trace-api-manual-tracing) for details.    |

Also check out the functions defined within `Datadog::Tracing` in our [public API](https://www.rubydoc.info/gems/ddtrace/) for more details on their usage.

<h3 id="1.0-trace-api-removed-tracer">Removed `Datadog.tracer`</h3>

Many of the functions accessed directly through `Datadog.tracer` have been moved to `Datadog::Tracing` instead.

<h3 id="1.0-trace-api-removed-context">Removed access to `Datadog::Context`</h3>

Direct usage of `Datadog::Context` has been removed. Previously, it was used to modify or access active trace state. Most use cases have been replaced by our [public trace API](https://www.rubydoc.info/gems/ddtrace/).

<h3 id="1.0-trace-api-manual-tracing">Manual tracing & trace model</h3>

> Manual tracing is now done through the [public API](https://www.rubydoc.info/gems/ddtrace/).

Reviewer: The rubydoc.info site hasn't yet been updated with 1.0.0.beta1 (probably only indexes non-beta versions?) but I think it would be nice to update this document after release with deeper links to specific classes/parts of the API. (So not for this PR, just a suggestion for a later improvement)

Use of `active_root_span` has been removed.

<h3 id="1.0-trace-api-distributed">Distributed tracing</h3>

Previously, distributed tracing required building new `Datadog::Context` objects, then replacing the context within the tracer.

> Instead, users utilize `TraceDigest` objects derived from a trace. This object represents the state of a trace. It can be used to propagate a trace across execution boundaries (processes, threads) or to continue a trace locally.

Author: I prefer shorter statements to conjunction here.

> Instead, users utilize `TraceDigest` objects derived from a trace. This object represents the state of a trace. It can be used to propagate a trace across execution boundaries (processes, threads) or to continue a trace locally.

Reviewer: Maybe then:

```suggestion
In 1.0, users need to use `TraceDigest` objects derived from a trace. `TraceDigest` represents the state of a trace. It can be used to propagate a trace across execution boundaries (processes, threads) or to continue a trace locally.
```

end
```

New in 1.0, it's also possible to explicitly assign a trace block to continue from a specific trace, rather than implicitly inherit an active context. This gives users fine-grained control in applications where multiple traces run concurrently in the same execution context:

```ruby
### New 1.0 ###
# Get trace continuation from active trace
trace_digest = Datadog::Tracing.active_trace.to_digest

# Continue a trace: explicit continuation
# Inherits trace properties from the trace digest
Datadog::Tracing.trace('my.job', continue_from: trace_digest) do |span, trace|
  trace.id == trace_digest.trace_id
end

# Continue a trace: explicit continuation (using #continue_from!)
Datadog::Tracing.continue_from!(trace_digest) do
  # Traces implicitly continue within the block
  Datadog::Tracing.trace('my.job') do |span, trace|
    trace.id == trace_digest.trace_id
  end
end
```

<h4 id="1.0-trace-api-distributed-http">Over HTTP</h4>

> *To propagate active trace to a remote service:*

Reviewer: Minor: *I don't think the italics here match the rest of the doc* 😄

| `Datadog::Worker`                                               | `Datadog::Core::Worker`                                                                 |
| `Datadog::Workers`                                              | `Datadog::Core::Workers`                                                                |

> <h3 id="1.0-appendix-breaking-changes">Breaking changes</h3>

Reviewer: ‼️ I think there was a search and replace accident in this whole section, with a lot of `:` becoming `%` and `$`.
